domain/operation/state/start.go

```diff
@@ -0,0 +1,617 @@
+}
+
+// StartExecOperationOnAllMachines creates an exec operation with tasks based
+// on the provided parameters on all machines.
+func (s *State) StartExecOperationOnAllMachines(ctx context.Context, args operation.ExecArgs) (operation.RunResult, error) {
+	db, err := s.DB(ctx)
+	if err != nil {
+		return operation.RunResult{}, errors.Capture(err)
+	}
+
+	var result operation.RunResult
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		machines, err := s.getAllMachines(ctx, tx)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		target := operation.Receivers{
+			Machines: machines,
+		}
+
+		result, err = s.startExecOperation(ctx, tx, target, args)
+		return errors.Capture(err)
+	})
+	if err != nil {
+		return operation.RunResult{}, errors.Errorf("starting exec operation on all machines: %w", err)
+	}
+
+	return result, nil
+}
+
+// StartActionOperation creates an action operation with tasks for various
+// units using the provided parameters.
+func (s *State) StartActionOperation(ctx context.Context,
+	target []operation.ActionReceiver,
+	args operation.TaskArgs) (operation.RunResult, error) {
+	db, err := s.DB(ctx)
+	if err != nil {
+		return operation.RunResult{}, errors.Capture(err)
+	}
+
+	var result operation.RunResult
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var err error
+		result, err = s.startActionOperation(ctx, tx, target, args)
+		return errors.Capture(err)
+	})
+	if err != nil {
+		return operation.RunResult{}, errors.Errorf("starting action operation: %w", err)
+	}
+
+	return result, nil
+}
+
+// startExecOperation implements the core logic for creating exec operations.
+func (s *State) startExecOperation(ctx context.Context, tx *sqlair.TX, target operation.Receivers, args operation.ExecArgs) (operation.RunResult, error) {
+	operationID, err := sequencestate.NextValue(ctx, s, tx, operation.OperationSequenceNamespace)
+	if err != nil {
+		return operation.RunResult{}, errors.Errorf("generating operation ID: %w", err)
+	}
+
+	operationUUID := internaluuid.MustNewUUID()
+	now := time.Now().UTC()
+
+	err = s.insertOperation(ctx, tx, insertOperation{
+		UUID:           operationUUID.String(),
+		OperationID:    strconv.Itoa(int(operationID)),
+		Summary:        fmt.Sprintf("exec %s", args.Command),
+		EnqueuedAt:     now,
+		Parallel:       args.Parallel,
+		ExecutionGroup: args.ExecutionGroup,
+	})
+	if err != nil {
+		return operation.RunResult{}, errors.Errorf("inserting operation: %w", err)
+	}
+
+	// Exec operations have command and timeout parameters.
+	err = s.insertOperationParameter(ctx, tx, operationUUID.String(), "command", args.Command)
+	if err != nil {
+		return operation.RunResult{}, errors.Errorf("inserting command parameter: %w", err)
+	}
+	err = s.insertOperationParameter(ctx, tx, operationUUID.String(), "timeout", args.Timeout.String())
+	if err != nil {
+		return operation.RunResult{}, errors.Errorf("inserting timeout parameter: %w", err)
+	}
```

Comment:
todo: this looks like a potato.  Based on the way the method is written, it can be a bulk insert.